runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPortableClientRunner.java

```diff
@@ -0,0 +1,250 @@
+    @Option(
+        name = "--driver-cmd",
+        required = true,
+        usage =
+            "Command that launches the Python driver program. "
+                + "(The job service endpoint will be appended as --job_endpoint=localhost:<port>.)")
+    private String driverCmd;
+  }
+
+  private static EntryPointConfiguration parseArgs(String[] args) {
+    EntryPointConfiguration configuration = new EntryPointConfiguration();
+    CmdLineParser parser = new CmdLineParser(configuration);
+    try {
+      parser.parseArgument(args);
+    } catch (CmdLineException e) {
+      LOG.error("Unable to parse command line arguments.", e);
+      parser.printUsage(System.err);
+      throw new IllegalArgumentException("Unable to parse command line arguments.", e);
+    }
+    return configuration;
+  }
+
+  private void startJobService() throws Exception {
+    jobInvokerFactory = new DetachedJobInvokerFactory();
+    driver =
+        FlinkJobServerDriver.fromConfig(
+            FlinkJobServerDriver.fromParams(
+                new String[] {"--job-port=" + jobPort, "--artifact-port=0", "--expansion-port=0"}),
+            jobInvokerFactory);
+    driverThread = new Thread(driver);
+    driverThread.start();
+
+    Duration timeout = Duration.ofSeconds(30);
+    Deadline deadline = Deadline.fromNow(timeout);
+    while (driver.getJobServerUrl() == null && deadline.hasTimeLeft()) {
+      try {
+        Thread.sleep(500);
+      } catch (InterruptedException interruptEx) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(interruptEx);
+      }
+    }
+
+    if (!driverThread.isAlive()) {
+      throw new IllegalStateException("Job service thread is not alive");
+    }
+
+    if (driver.getJobServerUrl() == null) {
+      String msg = String.format("Timeout of %s waiting for job service to start.", deadline);
+      throw new TimeoutException(msg);
+    }
+  }
+
+  private void runDriverProgram() throws Exception {
+    ProcessManager processManager = ProcessManager.create();
+    String executable = "bash";
+    List<String> args =
+        ImmutableList.of(
+            "-c",
+            String.format("exec %s " + DRIVER_CMD_FLAGS, driverCmd, driver.getJobServerUrl()));
+    String processId = "client1";
+
+    try {
+      final ProcessManager.RunningProcess driverProcess =
+          processManager.startProcess(processId, executable, args, System.getenv());
+      driverProcess.isAliveOrThrow();
+      LOG.info("Started driver program");
+
+      // await effect of the driver program submitting the job
+      jobInvokerFactory.executeDetachedJob();
+    } catch (Exception e) {
+      try {
+        processManager.stopProcess(processId);
+      } catch (Exception processKillException) {
+        e.addSuppressed(processKillException);
+      }
+      throw e;
+    }
+  }
+
+  private void stopJobService() throws InterruptedException {
+    if (driver != null) {
+      driver.stop();
```

Comment:
This should named `jobServer` to separate it from the pipeline submission driver.